source/framework/vx_graph_verify.c

```diff
@@ -105,6 +104,67 @@ static vx_status ownGraphUpdateObjArrRefAfterKernetInit(vx_object_array exemplar
 static vx_status ownGraphUpdatePyramidRefAfterKernetInit(vx_pyramid exemplar, vx_pyramid ref);
 static vx_status ownGraphAddSingleDataReference(vx_graph graph, vx_reference ref, uint32_t prm_dir, uint32_t check);
 
+/* Validate graph parameters; two parameters should not point to the same node & index, and two parameters should not have the same initial reference set. */
+static vx_status ownGraphValidateParameters(vx_graph graph)
+{
+    vx_status status = (vx_status)VX_SUCCESS;
+    vx_uint32 param_idx = 0U;
+    for (param_idx = 1U; param_idx < graph->num_params; ++param_idx)
+    {
+        vx_uint32 nxt_idx = 0U;
+        vx_reference param_ref = graph->parameters[param_idx-1].node->parameters[graph->parameters[param_idx-1].index];
+
+        for (nxt_idx = param_idx; nxt_idx < graph->num_params; ++nxt_idx)
+        {
+            if (param_ref == graph->parameters[nxt_idx].node->parameters[graph->parameters[nxt_idx].index])
+            {
+                status =(vx_status)VX_ERROR_INVALID_PARAMETERS;
+                VX_PRINT(VX_ZONE_ERROR, "Invalid graph parameters: #%d and #%d both attached to reference \"%s\"\n", param_idx - 1, nxt_idx, param_ref->name);
+            }
+        }
+    }
+    return status;
+}
+
+/* Link graph parameters if not pipelining or streaming */
+static vx_status ownGraphLinkParameters(vx_graph graph)
+{
+    vx_uint32 p_index;
+    vx_status status = (vx_status)VX_SUCCESS;
+    for (p_index = 0; p_index < graph->num_params; ++p_index)
+    {
+        vx_node node = graph->parameters[p_index].node;
+        vx_uint32 index = graph->parameters[p_index].index;
+        vx_reference ref = node->parameters[index];
+        vx_uint32 node_idx;
+        graph->parameters[p_index].num_other = 0;
+        for (node_idx = 0; node_idx < graph->num_nodes;++node_idx)
+        {
+            vx_node this_node = graph->nodes[node_idx];
+            vx_uint32 this_index;
+            for (this_index = 0; this_index < this_node->kernel->signature.num_parameters; ++this_index)
+            {
+                if (this_node != node || this_index != index)
+                {
+                    if (this_node->parameters[this_index] == ref)
+                    {
+                        /* we have another occurrence of the reference in the graph, record it */
+                        graph->parameters[p_index].params_list[graph->parameters[p_index].num_other].node = this_node;
+                        graph->parameters[p_index].params_list[graph->parameters[p_index].num_other].index = this_index;
+                        graph->parameters[p_index].num_other++;
+                        if (graph->parameters[p_index].num_other >= TIVX_GRAPH_MAX_PARAM_REFS)
+                        {
+                            VX_PRINT(VX_ZONE_ERROR, "Too many linked references for graph parameter %d, increase TIVX_GRAPH_MAX_PARAM_REFS\n", p_index);
```

Comment:
A few comments here:
1. Isn't this error premature?  Lets say TIVX_GRAPH_MAX_PARAM_REFS is set to 1.  It should be valid to hit this condition once, but not again.  However, the way this is written, the error is returned even the first time this is hit (since num_other gets incremented to 1, and this condition shows it is >=, and hits the error.  I suggest this check to be done at the beginning of this "if" section before the assignments.
2. Also, when this error is hit, shouldn't we break from this for loop, and the one outside of it so that we don't hit this condition multiple times per p_index?

Reply:
yes you're right. I modified the check